src/cephadm/cephadm.py

```diff
@@ -937,6 +938,136 @@ def get_tcmu_runner_container(self):
         tcmu_container.cname = self.get_container_name(desc='tcmu')
         return tcmu_container
 
+
+##################################
+
+
+class CephNvmeof(object):
+    """Defines a Ceph-Nvmeof container"""
+
+    daemon_type = 'nvmeof'
+    required_files = ['ceph-nvmeof.conf']
+    default_image = DEFAULT_NVMEOF_IMAGE
+
+    def __init__(self,
+                 ctx,
+                 fsid,
+                 daemon_id,
+                 config_json,
+                 image=DEFAULT_IMAGE):
+        # type: (CephadmContext, str, Union[int, str], Dict, str) -> None
+        self.ctx = ctx
+        self.fsid = fsid
+        self.daemon_id = daemon_id
+        self.image = image
+
+        # config-json options
+        self.files = dict_get(config_json, 'files', {})
+
+        # validate the supplied args
+        self.validate()
+
+    @classmethod
+    def init(cls, ctx, fsid, daemon_id):
+        # type: (CephadmContext, str, Union[int, str]) -> CephNvmeof
+        return cls(ctx, fsid, daemon_id,
+                   get_parm(ctx.config_json), ctx.image)
+
+    @staticmethod
+    def get_container_mounts(data_dir):
+        # type: (str, str) -> Dict[str, str]
+        mounts = dict()
+        mounts[os.path.join(data_dir, 'config')] = '/etc/ceph/ceph.conf:z'
+        #mounts[os.path.join(data_dir, 'keyring')] = '/etc/ceph/keyring:z'
+        mounts['/etc/ceph/ceph.client.admin.keyring'] = '/etc/ceph/keyring:z' # TODO(redo): FIXME
+        mounts[os.path.join(data_dir, 'ceph-nvmeof.conf')] = '/src/ceph-nvmeof.conf:z'
+        mounts[os.path.join(data_dir, 'configfs')] = '/sys/kernel/config'
+        #mounts[log_dir] = '/var/log:z' # TODO(redo): would we need a logdir?
+        mounts['/dev'] = '/dev'
+        return mounts
+
+    @staticmethod
+    def get_container_binds():
+        # type: () -> List[List[str]]
+        binds = []
+        lib_modules = ['type=bind',
+                       'source=/lib/modules',
+                       'destination=/lib/modules',
+                       'ro=true']
+        binds.append(lib_modules)
+        return binds
+
+    @staticmethod
+    def get_version(ctx, container_id):
+        # type: (CephadmContext, str) -> Optional[str]
+        #TODO(redo): get the right version of the nvmeof package or container
+        #version = None
+        #out, err, code = call(ctx,
+        #                      [ctx.container_engine.path, 'exec', container_id,
+        #                       '/usr/bin/python3', '-c', "import pkg_resources; print(pkg_resources.require('ceph_nvmeof')[0].version)"],
+        #                      verbosity=CallVerbosity.QUIET)
+        #if code == 0:
+        #    version = out.strip()
+        return "unknown"
```

Comment:
```suggestion
        # type: (CephadmContext, str) -> Optional[str]
        out, err, ret = call_throws(ctx, [
            ctx.container_engine.path, 'inspect',
            '--format', '{{index .Config.Labels "io.ceph.version"}}',
            ctx.image])
        version = None
        if ret == 0:
            version = out.strip()
        return version
```
The image version label is defined [here](https://github.com/epuertat/ceph-nvmeof/blob/80ba587e13e22d4723092d3b2b3b2852e53210e0/Dockerfile#L59).
@epuertat, does it make sense to fetch version from the image label?

Reply:
Great idea! I'm wondering if we should keep `io.ceph.version` to save the built-in Ceph (`rados` & `librbd`) versions used by spdk & nvmeof, and maybe add a `io.ceph.nvmeof.version` or `io.ceph.component.version`? What do you think @baum ?

I was still trying to figure out the best way to manage package versions & releases (nothing urgent, since we haven't released anything yet, but given we're talking about versions...):
* I removed the version from  `control/__init__.py` `__version__` (which is the most standard way to handle versions) and put that to `.env`, so that `docker-compose` commands don't need any runtime dynamic variable.
* Reading a `.env` file can be easily done with [`python-dotenv`](https://pypi.org/project/python-dotenv/) (in fact that's what `docker-compose` uses to read .env files).
* The Python Packager we're using (PDM) allows the following ways to [dynamically define the Python version](https://pdm-backend.fming.dev/metadata/#dynamic-project-version):
  * Reading that from `git` is something that I like 'cause Github makes it easy to specify tags/release.
  * Using a specific call (e.g.: `os.getenv("NVMEOF_VERSION")`).
  * It also allows to save the dynamic version to a text or a Python file.

What's your view?

Reply:
@epuertat @baum thank you very much for helping with this. Please, can you provide the definitive code to get the version? is it the code provided by @baum?